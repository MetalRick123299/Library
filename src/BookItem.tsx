import React, { useRef, useState } from 'react';
import clsx from 'clsx';
import { ChevronDownIcon, ChevronUpIcon } from '@heroicons/react/20/solid/esm';

export type IBookItem = {
  title: string;
  author: string;
  pagesRead: number;
  totalPages: number;
};

// TODOs
// Add Fast Book Page movement
// remove selecting outside elements while holding / spam clicking

export default function BookItem({ book }: { book: IBookItem }) {
  const [bookDetails, setBookDetails] = useState(book);
  const { title, author, pagesRead, totalPages } = bookDetails;
  const varyRef = useRef<NodeJS.Timer | null>(null);

  const startVaryPagesRead = (dir: 'down' | 'up') => {
    if (varyRef.current) return;

    if (dir === 'up') {
      varyRef.current = setInterval(() => {
        if (
          bookDetails.pagesRead + 1 < 0 ||
          bookDetails.pagesRead + 1 > bookDetails.totalPages
        )
          return;

        setBookDetails((prev) => ({ ...prev, pagesRead: prev.pagesRead + 1 }));
      }, 75);
    } else {
      varyRef.current = setInterval(() => {
        if (
          bookDetails.pagesRead - 1 < 0 ||
          bookDetails.pagesRead - 1 > bookDetails.totalPages
        )
          return;

        setBookDetails((prev) => ({ ...prev, pagesRead: prev.pagesRead - 1 }));
      }, 75);
    }
  };

  const stopVaryPagesRead = () => {
    if (varyRef.current) {
      clearInterval(varyRef.current);
      varyRef.current = null;

      if (bookDetails.pagesRead / book.totalPages > 1) {
        setBookDetails((prev) => ({ ...prev, pagesRead: prev.totalPages }));
      }

      // Add API Call to PUT new PageRead
    }
  };

  return (
    <div className="bg-primary-item flex flex-col items-center gap-3 p-5 rounded-xl w-64 text-2xl">
      <h2 className="">{title}</h2>
      <span className="">{author}</span>
      <span className="">
        {pagesRead} / {totalPages}
      </span>

      <div className="flex w-full items-center justify-between">
        <ChevronDownIcon
          className="h-8 cursor-pointer transition-all hover:scale-125 active:scale-95"
          onMouseDown={() => startVaryPagesRead('down')}
          onMouseUp={() => stopVaryPagesRead()}
          onMouseLeave={() => stopVaryPagesRead()}
        />

        <div className="w-8/12 rounded-lg border-2 border-primary-bg text-center relative z-[1]">
          {((pagesRead / totalPages) * 100).toFixed(2)}%
          <div
            className="h-full bg-primary-bg absolute inset-0 -z-[1] transition-all max-w-full"
            style={{
              width: `${((pagesRead / totalPages) * 100).toFixed(2)}%`,
            }}
          />
        </div>

        <ChevronUpIcon
          className="h-8 cursor-pointer transition-all hover:scale-125 active:scale-95 z-[1]"
          onMouseDown={() => startVaryPagesRead('up')}
          onMouseUp={() => stopVaryPagesRead()}
          onMouseLeave={() => stopVaryPagesRead()}
        />
      </div>

      <button type="button" className="bg-red-500 self-stretch rounded-lg p-1">
        Delete
      </button>
    </div>
  );
}
