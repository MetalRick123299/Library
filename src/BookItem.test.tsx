import { render, screen } from '@testing-library/react';
import BookItem, { IBookItem } from './BookItem';

describe('Book Item', () => {
  const book: IBookItem = {
    title: 'Harry Potter',
    author: 'J.K Rowling',
    pagesRead: 45,
    totalPages: 586,
  };
  const book2: IBookItem = {
    title: 'The Great Gatsby',
    author: 'F. Scott Fitzgerald',
    pagesRead: 104,
    totalPages: 208,
  };
  const book3: IBookItem = {
    title: 'The Lightning Thief',
    author: 'Rick Riordan',
    pagesRead: 372,
    totalPages: 375,
  };
  const bookArray = [book, book2, book3];

  it('BookItem Structure', () => {
    render(<BookItem book={book} />);

    const title = screen.getByRole('heading', { level: 2, name: book.title });
    const author = screen.getByText(book.author);
    const pagesElement = screen.getByText(
      `${book.pagesRead} / ${book.totalPages}`
    );
    const progress = screen.getByText(
      `${parseFloat((book.pagesRead / book.totalPages).toFixed(4)) * 100}%`
    );
    const deleteButton = screen.getByRole('button', { name: 'Delete' });

    expect(title.innerHTML).toBe(book.title);
    expect(author.innerHTML).toBe(book.author);
    expect(pagesElement.innerHTML).toBe(
      `${book.pagesRead} / ${book.totalPages}`
    );
    expect(progress).toBe(progress);
    expect(deleteButton).toBe(deleteButton);
  });

  it('Multiple books', () => {
    bookArray.forEach((currBook) => {
      render(<BookItem book={currBook} />);

      const title = screen.getByRole('heading', {
        level: 2,
        name: currBook.title,
      });
      const author = screen.getByText(currBook.author);
      const pagesElement = screen.getByText(
        `${currBook.pagesRead} / ${currBook.totalPages}`
      );
      const progress = screen.getByText(
        `${((currBook.pagesRead / currBook.totalPages) * 100).toFixed(2)}%`
      );

      expect(title.innerHTML).toBe(currBook.title);
      expect(author.innerHTML).toBe(currBook.author);
      expect(pagesElement.innerHTML).toContain(
        `${currBook.pagesRead} / ${currBook.totalPages}`
      );
      expect(progress).toBe(progress);
    });
  });

  it('SnapShot', () => {
    const bookItem = render(<BookItem book={book} />);
    expect(bookItem).toMatchSnapshot();
  });

  it('Progress Bar Shows % of current book read', () => {
    bookArray.forEach((currBook) => {
      render(<BookItem book={currBook} />);

      const percentRead = `${(
        (currBook.pagesRead / currBook.totalPages) *
        100
      ).toFixed(2)}%`;
      const progressbar = screen.getByText(percentRead);

      expect(progressbar.innerHTML).toContain(percentRead);
      expect(progressbar.innerHTML).toContain(`style="width: ${percentRead};`);
    });
  });

  // Do => 11/22/2022
  it.todo('User allowed to quickly increases Pages Read', () => {});
  it.todo('Progress Bar Changes as Pages Read / Total Pages Change', () => {});

  // Do => 11/23/2022 - 11/24/2022
  it.todo('Delete Button deletes todo', () => {});
  it.todo('Modal for Edit Book', () => {});
  it.todo('Edit Todo Works', () => {});
});
