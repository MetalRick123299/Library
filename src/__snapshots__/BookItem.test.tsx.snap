// Vitest Snapshot v1

exports[`Book Item > SnapShot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="bg-primary-item flex flex-col items-center gap-3 p-5 rounded-xl w-64 text-2xl"
      >
        <h2
          class=""
        >
          Harry Potter
        </h2>
        <span
          class=""
        >
          J.K Rowling
        </span>
        <span
          class=""
        >
          45
           / 
          586
        </span>
        <div
          class="flex w-full items-center justify-between"
        >
          <svg
            aria-hidden="true"
            class="h-8 cursor-pointer transition-all hover:scale-125 active:scale-95"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M5.23 7.21a.75.75 0 011.06.02L10 11.168l3.71-3.938a.75.75 0 111.08 1.04l-4.25 4.5a.75.75 0 01-1.08 0l-4.25-4.5a.75.75 0 01.02-1.06z"
              fill-rule="evenodd"
            />
          </svg>
          <div
            class="w-8/12 rounded-lg border-2 border-primary-bg text-center relative z-[1]"
          >
            7.68
            %
            <div
              class="h-full bg-primary-bg absolute inset-0 -z-[1] transition-all max-w-full"
              style="width: 7.68%;"
            />
          </div>
          <svg
            aria-hidden="true"
            class="h-8 cursor-pointer transition-all hover:scale-125 active:scale-95 z-[1]"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M14.77 12.79a.75.75 0 01-1.06-.02L10 8.832 6.29 12.77a.75.75 0 11-1.08-1.04l4.25-4.5a.75.75 0 011.08 0l4.25 4.5a.75.75 0 01-.02 1.06z"
              fill-rule="evenodd"
            />
          </svg>
        </div>
        <button
          class="bg-red-500 self-stretch rounded-lg p-1"
          type="button"
        >
          Delete
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="bg-primary-item flex flex-col items-center gap-3 p-5 rounded-xl w-64 text-2xl"
    >
      <h2
        class=""
      >
        Harry Potter
      </h2>
      <span
        class=""
      >
        J.K Rowling
      </span>
      <span
        class=""
      >
        45
         / 
        586
      </span>
      <div
        class="flex w-full items-center justify-between"
      >
        <svg
          aria-hidden="true"
          class="h-8 cursor-pointer transition-all hover:scale-125 active:scale-95"
          fill="currentColor"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M5.23 7.21a.75.75 0 011.06.02L10 11.168l3.71-3.938a.75.75 0 111.08 1.04l-4.25 4.5a.75.75 0 01-1.08 0l-4.25-4.5a.75.75 0 01.02-1.06z"
            fill-rule="evenodd"
          />
        </svg>
        <div
          class="w-8/12 rounded-lg border-2 border-primary-bg text-center relative z-[1]"
        >
          7.68
          %
          <div
            class="h-full bg-primary-bg absolute inset-0 -z-[1] transition-all max-w-full"
            style="width: 7.68%;"
          />
        </div>
        <svg
          aria-hidden="true"
          class="h-8 cursor-pointer transition-all hover:scale-125 active:scale-95 z-[1]"
          fill="currentColor"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M14.77 12.79a.75.75 0 01-1.06-.02L10 8.832 6.29 12.77a.75.75 0 11-1.08-1.04l4.25-4.5a.75.75 0 011.08 0l4.25 4.5a.75.75 0 01-.02 1.06z"
            fill-rule="evenodd"
          />
        </svg>
      </div>
      <button
        class="bg-red-500 self-stretch rounded-lg p-1"
        type="button"
      >
        Delete
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
